mondrian/src/main/java/mondrian/rolap/SqlMemberSource.java

```diff
@@ -420,6 +425,82 @@ private List<RolapMember> getMembers(DataSource dataSource) {
         }
     }
 
+    private int countMemberChildren(RolapMember parentMember, List<RolapMember> children, MemberChildrenConstraint constraint) {
+
+        Pair<String, List<SqlStatement.Type>> pair;
+        SqlQuery query = null;
+        boolean parentChild;
+        final RolapLevel parentLevel = parentMember.getLevel();
+        RolapLevel childLevel;
+        if ( parentLevel.isParentChild() ) {
+            query = makeChildMemberSqlPCQuery( parentMember );
+            parentChild = true;
+            childLevel = parentLevel;
+        } else {
+            childLevel = (RolapLevel) parentLevel.getChildLevel();
+            if ( childLevel == null ) {
+                // member is at last level, so can have no children
+                return 0;
+            }
+            if ( childLevel.isParentChild() ) {
+                query = makeChildMemberSql_PCRootQuery( parentMember );
+                parentChild = true;
+            } else {
+               // query = makeChildMemberSqlQuery( parentMember, dataSource, constraint );
+                parentChild = false;
+                query = makeChildMemberSqlQuery( parentMember, dataSource, constraint );
+            }
+        }
+
+        SqlQuery newQuery =
+          SqlQuery.newQuery(
+            dataSource,
+            "while generating query to retrieve children of "
+              + "parent/child hierarchy member " + parentMember);
+
+        newQuery.addSelect("count(*)", null);
+        newQuery.addFrom( query, "X", false );
+
+
+        pair = newQuery.toSqlAndTypes();
+        final String sql = pair.left;
+
+        Integer result = null;
+
+        final List<SqlStatement.Type> types = pair.right;
+        SqlStatement stmt =
+          RolapUtil.executeQuery(
+            dataSource, sql, types, 0, 0,
+            new SqlStatement.StatementLocus(
+              Locus.peek().execution,
+              "SqlMemberSource.getMemberChildren",
+              "while building member cache",
+              SqlStatementEvent.Purpose.TUPLES, 0),
+            -1, -1, null);
+        try {
+
+            final List<SqlStatement.Accessor> accessors = stmt.getAccessors();
+            ResultSet resultSet = stmt.getResultSet();
+            while (resultSet.next()) {
```

Comment:
Why use a 'while' when it will be a one iteration only?
It could/should be an 'if'.